File: src-tauri/src/mert_worker.rs
//! Bridge to the MERT-95M feature-extraction python worker.
//!
//! Shells out to `mert_worker.py` against a full-mix audio file *and* the
//! demucs drum stem, producing two cached .npy files in a single Python
//! process. The model is loaded once per track — consumer-laptop friendly
//! — and the resulting caches feed the bar classifier (full mix) and the
//! n2n drum-onset preprocessor (drum stem).

use serde::Deserialize;
use std::path::{Path, PathBuf};
use std::process::Command;
use tauri::AppHandle;

use crate::python_env;

const WORKER_SOURCE: &str = include_str!("../python/mert_worker.py");
const WORKER_SCRIPT_NAME: &str = "mert_worker.py";

/// Result of a successful MERT extraction. Both caches written by one
/// Python invocation against the same loaded MERT-95M model.
#[derive(Debug, Clone)]
pub struct MertCache {
    pub fullmix_path: PathBuf,
    pub drum_path: PathBuf,
    #[allow(dead_code)]
    pub fullmix_frames: u64,
    #[allow(dead_code)]
    pub drum_frames: u64,
}

#[derive(Deserialize)]
struct WorkerResponse {
    fullmix_path: String,
    drum_path: String,
    fullmix_frames: u64,
    drum_frames: u64,
    #[allow(dead_code)]
    frames_per_second: u32,
    #[allow(dead_code)]
    layer: u32,
    #[allow(dead_code)]
    model_id: String,
}

pub fn compute_mert_cache(
    app: &AppHandle,
    fullmix_path: &Path,
    drum_path: &Path,
    out_fullmix: &Path,
    out_drum: &Path,
) -> Result<MertCache, String> {
    let python_path = python_env::ensure_python_env(app)?;
    let script_path = python_env::ensure_worker_script(app, WORKER_SCRIPT_NAME, WORKER_SOURCE)?;
    // The worker imports `n2n.infer.compute_mert_features` so it shares the
    // exact chunking parameters the training pipeline uses; ensure the n2n
    // resource dir is unpacked alongside the script and run with workdir =
    // script's parent so the import resolves.
    let _ = python_env::ensure_python_resource_dir(app, "n2n")?;
    let workdir = script_path
        .parent()
        .ok_or_else(|| "Worker script missing parent directory".to_string())?;

    for out in [out_fullmix, out_drum] {
        if let Some(parent) = out.parent() {
            std::fs::create_dir_all(parent).map_err(|e| {
                format!("Failed to create MERT cache dir {}: {e}", parent.display())
            })?;
        }
    }

    let mut cmd = Command::new(&python_path);
    crate::cmd_util::no_window(&mut cmd);
    let output = cmd
        .env("PYTHONUNBUFFERED", "1")
        .arg(&script_path)
        .arg("--fullmix")
        .arg(fullmix_path)
        .arg("--drum")
        .arg(drum_path)
        .arg("--out-fullmix")
        .arg(out_fullmix)
        .arg("--out-drum")
        .arg(out_drum)
        .current_dir(workdir)
        .output()
        .map_err(|e| format!("Failed to launch MERT worker: {e}"))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
        return Err(if stderr.is_empty() {
            "MERT worker exited unsuccessfully".to_string()
        } else {
            format!("MERT worker failed: {stderr}")
        });
    }

    let stdout = String::from_utf8(output.stdout)
        .map_err(|e| format!("MERT worker output was not valid UTF-8: {e}"))?;
    let payload: WorkerResponse = serde_json::from_str(stdout.trim())
        .map_err(|e| format!("Failed to parse MERT response '{}': {e}", stdout.trim()))?;

    Ok(MertCache {
        fullmix_path: PathBuf::from(payload.fullmix_path),
        drum_path: PathBuf::from(payload.drum_path),
        fullmix_frames: payload.fullmix_frames,
        drum_frames: payload.drum_frames,
    })
}

File: src-tauri/python/mert_worker.py
#!/usr/bin/env python3
"""MERT-95M feature-extraction worker.

Computes per-track MERT-v1-95M layer-7 hidden states from two inputs — the
full-mix audio and the demucs drum stem — and writes each to its own fp16
numpy cache. Both passes happen inside this single process so MERT-95M is
loaded into memory (and GPU/MPS) exactly once per track, which dominates
runtime on consumer laptops.

The full-mix cache feeds the bar classifier (per-bar slicing for the
22-tag windowed head). The drum-stem cache feeds the n2n drum-onset model
— v6+ checkpoints were trained on drum-isolated stems, so its MERT
conditioning has to be on the drum stem, not the full mix.

Algorithm: delegated to `n2n.infer.compute_mert_features` so chunking
parameters (window/overlap/crop) stay locked to the same defaults the
training pipeline and `python -m n2n.infer` use. Duplicating the math here
once cost us a ~60× drop in detected hats on PARTY 4 U after we noticed our
worker was using 60/30/5 while training used 30/15/3 — silent
distribution shift. Single source of truth from now on.

CLI:
    mert_worker.py --fullmix <wav> --drum <wav> \
        --out-fullmix <cache_path> --out-drum <cache_path>

Output (stdout, JSON):
    {
        "fullmix_path": "...",
        "drum_path": "...",
        "fullmix_frames": N,
        "drum_frames": M,
        "frames_per_second": 75,
        "layer": 7,
        "model_id": "m-a-p/MERT-v1-95M"
    }
"""

from __future__ import annotations

import argparse
import contextlib
import json
import pathlib
import sys


MERT_MODEL_ID = "m-a-p/MERT-v1-95M"
MERT_FRAMES_PER_SECOND = 75
MERT_LAYER = 7
# `n2n.infer.compute_mert_features` consumes audio at cfg["mel"]["sample_rate"]
# (the n2n training rate) and resamples to 24 kHz internally. Hard-coding here
# keeps us aligned with the bundled v12 checkpoint without parsing it.
N2N_TARGET_SAMPLE_RATE = 44_100


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--fullmix", type=pathlib.Path, required=True,
                   help="Path to the full-mix audio file.")
    p.add_argument("--drum", type=pathlib.Path, required=True,
                   help="Path to the demucs drum stem audio file.")
    p.add_argument("--out-fullmix", type=pathlib.Path, required=True,
                   help="Where to write the full-mix .npy cache.")
    p.add_argument("--out-drum", type=pathlib.Path, required=True,
                   help="Where to write the drum-stem .npy cache.")
    return p.parse_args()


def _atomic_write_npy(features, out_path: pathlib.Path) -> None:
    """Write `.npy` via tmp → rename so a killed process can't leave a
    truncated cache visible to consumers."""
    import numpy as np

    out_path.parent.mkdir(parents=True, exist_ok=True)
    tmp = out_path.with_suffix(out_path.suffix + ".tmp")
    with tmp.open("wb") as f:
        np.save(f, features.astype(np.float16), allow_pickle=False)
    tmp.replace(out_path)


def main() -> int:
    args = parse_args()
    for label, path in (("fullmix", args.fullmix), ("drum", args.drum)):
        if not path.exists():
            print(json.dumps({"error": f"{label} audio does not exist: {path}"}),
                  file=sys.stderr)
            return 1

    # Heavy imports + transformers logging redirected to stderr; stdout is
    # reserved for the final JSON payload.
    fullmix_frames = 0
    drum_frames = 0
    with contextlib.redirect_stdout(sys.stderr):
        try:
            import torch
            from transformers import AutoFeatureExtractor, AutoModel
            from n2n.infer import compute_mert_features, load_audio
        except Exception as exc:
            print(json.dumps({"error": f"Failed to import deps: {exc}"}), file=sys.stderr)
            return 1

        try:
            if torch.cuda.is_available():
                device = torch.device("cuda")
            elif torch.backends.mps.is_available():
                device = torch.device("mps")
            else:
                device = torch.device("cpu")

            fe = AutoFeatureExtractor.from_pretrained(MERT_MODEL_ID, trust_remote_code=True)
            model = AutoModel.from_pretrained(MERT_MODEL_ID, trust_remote_code=True).to(device).eval()
            for p in model.parameters():
                p.requires_grad_(False)

            # Minimal cfg surface used by `compute_mert_features`. Matches the
            # bundled n2n v12 checkpoint's relevant fields.
            cfg = {
                "data": {"mert_layer": MERT_LAYER},
                "mel": {"sample_rate": N2N_TARGET_SAMPLE_RATE},
            }

            for label, audio_path, out_path in (
                ("fullmix", args.fullmix, args.out_fullmix),
                ("drum", args.drum, args.out_drum),
            ):
                # `load_audio` from n2n.infer matches what `n2n.infer.main()`
                # does on the training machine: load + mono-mix + resample to
                # the n2n target SR; then compute_mert_features handles the
                # second resample to 24 kHz internally.
                audio = load_audio(audio_path, target_sr=N2N_TARGET_SAMPLE_RATE)
                features = compute_mert_features(
                    audio,
                    cfg,
                    device,
                    mert_model=model,
                    feature_extractor=fe,
                )  # (T, 768) fp32 on CPU — defaults: 30 s / 15 s / 3 s
                features_np = features.cpu().numpy()
                _atomic_write_npy(features_np, out_path)
                if label == "fullmix":
                    fullmix_frames = int(features_np.shape[0])
                else:
                    drum_frames = int(features_np.shape[0])
        except Exception as exc:
            print(json.dumps({"error": str(exc)}), file=sys.stderr)
            return 1

    sys.stdout.write(json.dumps({
        "fullmix_path": str(args.out_fullmix),
        "drum_path": str(args.out_drum),
        "fullmix_frames": fullmix_frames,
        "drum_frames": drum_frames,
        "frames_per_second": MERT_FRAMES_PER_SECOND,
        "layer": MERT_LAYER,
        "model_id": MERT_MODEL_ID,
    }))
    sys.stdout.flush()
    return 0


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())
